src/main/java/com/epam/aidial/core/service/ApplicationService.java

```diff
@@ -0,0 +1,282 @@
+                                                                   Application application, boolean preserveReference) {
+        prepareApplication(resource, etag, application, preserveReference);
+
+        if (application.getFunction() != null) {
+            throw new HttpException(HttpStatus.CONFLICT, "Currently not supported");
+        }
+
+        ResourceItemMetadata meta = resourceService.computeResource(resource, etag, json -> {
+            Application existing = ProxyUtil.convertToObject(json, Application.class, true);
+            Application.Function function = application.getFunction();
+
+            if (function != null) {
+                function.setStatus(Application.Function.Status.CREATED);
+                function.setState(new Application.Function.State());
+            }
+
+            if (existing != null && existing.getFunction() != null) {
+                Application.Function existingFunction = existing.getFunction();
+
+                if (function == null && existingFunction.getStatus() != Application.Function.Status.CREATED) {
+                    throw new HttpException(HttpStatus.CONFLICT, "The previous application must be deleted");
+                }
+
+                if (function != null) {
+                    function.setTargetFolder(existing.getFunction().getTargetFolder());
+                    function.setStatus(existing.getFunction().getStatus());
+                    function.setState(existing.getFunction().getState());
+                }
+            }
+
+            return ProxyUtil.convertToString(application, true);
+        });
+
+        return Pair.of(meta, application);
+    }
+
+    public void deleteApplication(ResourceDescription resource, EtagHeader etag) {
+        if (resource.isFolder() || resource.getType() != ResourceType.APPLICATION) {
+            throw new IllegalArgumentException("Invalid application url: " + resource.getUrl());
+        }
+
+        resourceService.computeResource(resource, etag, json -> {
+            Application application = ProxyUtil.convertToObject(json, Application.class, true);
+            if (application == null) {
+                throw new HttpException(HttpStatus.NOT_FOUND, "Application is not found: " + resource.getUrl());
+            }
+
+            Application.Function function = application.getFunction();
+            if (function != null && (function.getStatus() == Application.Function.Status.STARTING
+                            || function.getStatus() == Application.Function.Status.STARTED
+                            || function.getStatus() == Application.Function.Status.STOPPING)) {
+                throw new HttpException(HttpStatus.NOT_FOUND, "Application must be stopped: " + resource.getUrl());
+            }
+
+            return null;
+        });
+    }
+
+    public void copyApplication(ResourceDescription source, ResourceDescription description, boolean overwrite) {
```

Comment:
Let's rename the 2nd arg to `destination`

Reply:
renamed